Objects/longobject.c

```diff
@@ -15,35 +15,78 @@
 #include <float.h>
 #include <stddef.h>
 #include <stdlib.h>               // abs()
+#include <stdbool.h>
+#include <string.h>
+#include <limits.h>
 
 #include "clinic/longobject.c.h"
 /*[clinic input]
 class int "PyObject *" "&PyLong_Type"
 [clinic start generated code]*/
 /*[clinic end generated code: output=da39a3ee5e6b4b0d input=ec0275e3422a36e3]*/
 
-/* Is this PyLong of size 1, 0 or -1? */
-#define IS_MEDIUM_VALUE(x) (((size_t)Py_SIZE(x)) + 1U < 3U)
+#define IS_SMALL_INT(ival) (-_PY_NSMALLNEGINTS <= (ival) && (ival) < _PY_NSMALLPOSINTS)
+#define IS_SMALL_UINT(ival) ((ival) < _PY_NSMALLPOSINTS)
 
-/* convert a PyLong of size 1, 0 or -1 to a C integer */
-static inline stwodigits
-medium_value(PyLongObject *x)
+/* FIXME(lpereira): These predicates/utility functions can be marked with __attribute__((pure)) so that,
+ * when inlined, part of their computation can be avoided if called repeatedly with the same argument. */
+Py_LOCAL_INLINE(bool)
+fits_in_size_field(Py_ssize_t size)
 {
-    assert(IS_MEDIUM_VALUE(x));
-    return ((stwodigits)Py_SIZE(x)) * x->ob_digit[0];
+    /* 1 bit for the sign bit, 1 bit for the "is inlined or bignum" flag */
+    return (Py_size_t)size < 1UL<<(sizeof(size) * CHAR_BIT - 2);
 }
 
-#define IS_SMALL_INT(ival) (-_PY_NSMALLNEGINTS <= (ival) && (ival) < _PY_NSMALLPOSINTS)
-#define IS_SMALL_UINT(ival) ((ival) < _PY_NSMALLPOSINTS)
+Py_LOCAL_INLINE(bool)
+as_inlined_int(const PyLongObject *x, Py_ssize_t *value)
+{
+    *value = (Py_ssize_t)((Py_size_t)x->ob_size >> 1UL);
+    return x->ob_size & 1UL;
+}
 
-static inline int is_medium_int(stwodigits x)
+Py_LOCAL_INLINE(bool)
+is_zero(const PyLongObject *x)
 {
-    /* Take care that we are comparing unsigned values. */
-    twodigits x_plus_mask = ((twodigits)x) + PyLong_MASK;
-    return x_plus_mask < ((twodigits)PyLong_MASK) + PyLong_BASE;
+    Py_ssize_t value;
+    (void)as_inlined_int(x, &value);
+    return value == 0;
 }
 
-static PyObject *
+Py_LOCAL_INLINE(bool)
+is_negative(const PyLongObject *x)
+{
+    Py_ssize_t value;
+    (void)as_inlined_int(x, &value);
+    return value < 0;
+}
+
+Py_LOCAL_INLINE(Py_ssize_t)
+encode_value(Py_ssize_t value)
+{
+    assert(fits_in_size_field(value));
+    return (Py_ssize_t)(((Py_size_t)value << 1) | 1UL);
+}
+
+Py_LOCAL_INLINE(Py_ssize_t)
+encode_size(Py_ssize_t value)
+{
+    assert(!fits_in_size_field(value));
```

Comment:
No, this is used to encode the number of digits for big nums.  It's not supposed to be used when we're encoding the value that will be used in `ob_size` for inlined ints (they have the LSB set).

Maybe the function has to have a better name to avoid this confusion.